### PR TITLE
ci: use a separate workflow for codecov upload

### PR DESCRIPTION
**Pull request by superlopuh**
My understanding is that actions that are triggered by the workflow_run action are run with secrets enabled.

### CODE REVIEW

File: .github/workflows/ci-codecov.yml [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    - name: Upload coverage to Codecov
+      uses: Wandalen/wretry.action@v1
+      with:
+        action: codecov/codecov-action@v4
+        attempt_delay: 10000
+        attempt_limit: 10
+        with: |
+          fail_ci_if_error: true
+          verbose: true
+          files: coverage.xml
```
Comment (superlopuh) [inline .github/workflows/ci-codecov.yml:27]: I'm not entirely sure how this file gets passed around, I've done a bit of googling and it looks like it should "just work"

**Review by georgebisbas — APPROVED** [review]
We can merge and then rebase some PR to see how it behaves

File: .github/workflows/ci-codecov.yml (hunk 1 above)
Comment (PapyChacal) [reply .github/workflows/ci-codecov.yml:27]: I don't think so, the usual mechanism I know of is artifacts:
https://docs.github.com/en/actions/using-workflows/storing-workflow-data-as-artifacts

File: .github/workflows/ci-codecov.yml [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    types: [completed]
+
+jobs:
+  upload-codecov:
+    runs-on: ubuntu-20.04
+    if: ${{ github.event.workflow_run.conclusion == 'success' }}
+    steps:
+    - uses: actions/checkout@v4
+      with:
+        path: xdsl
```
Comment (PapyChacal) [inline .github/workflows/ci-codecov.yml:16]: I don't think we need to checkout xDSL to upload the report

**Comment by PapyChacal** [comment]
> It looks like the second workflow has to be in main for it to get triggered. Maybe the best thing to do would be to merge this PR (if it looks fine to everyone) and retry with a forked repo PR to see if it works.

I wouldn't block this but I want to give the alternative, as we usually religiously keep main intact: Just merge this on a fork's default branch, and it's good enough for GitHubs' criterion.
I'm guessing it will have issues with secrets not set on the fork and what not; but is good enough for testing the trigger mechanism :slightly_smiling_face:

**Comment by superlopuh** [comment]
IDK, I feel like this is worth trying anyway, I'm hoping this will fix our own dependabot issues also

**Pull request merged**